internal/provider/configure_clients.go

```diff
@@ -0,0 +1,310 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse API url: %v", err.Error())
+	}
+	apiPath, err := url.JoinPath(u.Path, "api")
+	if err != nil {
+		return nil, fmt.Errorf("failed to join API path: %v", err.Error())
+	}
+
+	userInfo, orgID, APIKey, err := parseAuth(providerConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg := goapi.TransportConfig{
+		Host:             u.Host,
+		BasePath:         apiPath,
+		Schemes:          []string{u.Scheme},
+		NumRetries:       int(providerConfig.Retries.ValueInt64()),
+		RetryTimeout:     time.Second * time.Duration(providerConfig.RetryWait.ValueInt64()),
+		RetryStatusCodes: setToStringArray(providerConfig.RetryStatusCodes.Elements()),
+		TLSConfig:        tlsClientConfig,
+		BasicAuth:        userInfo,
+		OrgID:            orgID,
+		APIKey:           APIKey,
+	}
+
+	if cfg.HTTPHeaders, err = getHTTPHeadersMap(providerConfig); err != nil {
+		return nil, err
+	}
+
+	return goapi.NewHTTPClientWithConfig(strfmt.Default, &cfg), nil
+}
+
+func createMLClient(url string, grafanaCfg *gapi.Config) (*mlapi.Client, error) {
+	mlcfg := mlapi.Config{
+		BasicAuth:   grafanaCfg.BasicAuth,
+		BearerToken: grafanaCfg.APIKey,
+		Client:      grafanaCfg.Client,
+		NumRetries:  grafanaCfg.NumRetries,
+	}
+	mlURL := url
+	if !strings.HasSuffix(mlURL, "/") {
+		mlURL += "/"
+	}
+	mlURL += "api/plugins/grafana-ml-app/resources"
+	mlclient, err := mlapi.New(mlURL, mlcfg)
+	if err != nil {
+		return nil, err
+	}
+	return mlclient, nil
+}
+
+func createCloudClient(providerConfig frameworkProviderConfig) (*gapi.Client, error) {
+	cfg := gapi.Config{
+		APIKey:       providerConfig.CloudAPIKey.ValueString(),
+		NumRetries:   int(providerConfig.Retries.ValueInt64()),
+		RetryTimeout: time.Second * time.Duration(providerConfig.RetryWait.ValueInt64()),
+	}
+
+	var err error
+	if cfg.HTTPHeaders, err = getHTTPHeadersMap(providerConfig); err != nil {
+		return nil, err
+	}
+
+	return gapi.New(providerConfig.CloudAPIURL.ValueString(), cfg)
+}
+
+func createOnCallClient(providerConfig frameworkProviderConfig) (*onCallAPI.Client, error) {
+	return onCallAPI.New(providerConfig.OncallURL.ValueString(), providerConfig.OncallAccessToken.ValueString())
+}
+
+// Sets a custom HTTP Header on all requests coming from the Grafana Terraform Provider to Grafana-Terraform-Provider: true
+// in addition to any headers set within the `http_headers` field or the `GRAFANA_HTTP_HEADERS` environment variable
+func getHTTPHeadersMap(providerConfig frameworkProviderConfig) (map[string]string, error) {
+	headers := map[string]string{"Grafana-Terraform-Provider": "true"}
+	for k, v := range providerConfig.HTTPHeaders.Elements() {
+		if vString, ok := v.(types.String); ok {
+			headers[k] = vString.ValueString()
+		} else {
+			return nil, fmt.Errorf("invalid header value for %s: %v", k, v)
+		}
+	}
+
+	return headers, nil
+}
+
+func createTempFileIfLiteral(value string) (path string, tempFile bool, err error) {
```

Comment:
You can remove `tempFile bool` and check if the path is empty. In that case, as the way that you used this function, you should return an empty string instead of `value`.